libs/SalesforceOAuth/SalesforceOAuth/Classes/SFOAuthSessionRefresher.m

```diff
@@ -0,0 +1,122 @@
+/*
+ Copyright (c) 2015, salesforce.com, inc. All rights reserved.
+ 
+ Redistribution and use of this software in source and binary forms, with or without modification,
+ are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this list of conditions
+ and the following disclaimer.
+ * Redistributions in binary form must reproduce the above copyright notice, this list of
+ conditions and the following disclaimer in the documentation and/or other materials provided
+ with the distribution.
+ * Neither the name of salesforce.com, inc. nor the names of its contributors may be used to
+ endorse or promote products derived from this software without specific prior written
+ permission of salesforce.com, inc.
+ 
+ THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
+ IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+ FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+ CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+ WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "SFOAuthSessionRefresher+Internal.h"
+#import "SFOAuthCredentials.h"
+
+@implementation SFOAuthSessionRefresher
+
+- (instancetype)initWithCredentials:(SFOAuthCredentials *)credentials {
+    self = [super init];
+    if (self) {
+        self.coordinator = [[SFOAuthCoordinator alloc] initWithCredentials:credentials];
+        self.coordinator.delegate = self;
+    }
+    return self;
+}
+
+- (instancetype)init {
+    return [self initWithCredentials:nil];
+}
+
+- (void)dealloc {
+    if (self.coordinator.isAuthenticating) {
+        [self.coordinator stopAuthentication];
+    }
+    self.coordinator.delegate = nil;
+}
+
+- (void)refreshSessionWithCompletion:(void (^)(SFOAuthCredentials *))completionBlock error:(void (^)(NSError *))errorBlock {
+    self.completionBlock = completionBlock;
+    self.errorBlock = errorBlock;
+    if (self.coordinator.credentials.instanceUrl == nil) {
```

Comment:
Yeah.  It's basically just a sanity check on credentials data that's necessary to facilitate a refresh.